snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+	continuousStragglingPeriod time.Duration
+
+	// previousStragglerCheckTime is the last time we checked whether
+	// our block height is behind the rest of the network
+	previousStragglerCheckTime time.Time
+
+	// prevSnapshot is the snapshot from a past iteration.
+	prevSnapshot snapshot
+}
+
+func newStragglerDetector(
+	getTime func() time.Time,
+	log logging.Logger,
+	minConfirmationThreshold float64,
+	lastAccepted func() (ids.ID, uint64),
+	connectedValidators func() set.Set[ids.NodeWeight],
+	connectedPercent func() float64,
+	processing func(id ids.ID) bool,
+	lastAcceptedByNodeID func(ids.NodeID) (ids.ID, bool),
+) *stragglerDetector {
+	sd := &stragglerDetector{
+		stragglerDetectorConfig: stragglerDetectorConfig{
+			lastAccepted:              dropHeight(lastAccepted),
+			processing:                processing,
+			minStragglerCheckInterval: minStragglerCheckInterval,
+			log:                       log,
+			connectedValidators:       connectedValidators,
+			connectedPercent:          connectedPercent,
+			minConfirmationThreshold:  minConfirmationThreshold,
+			lastAcceptedByNodeID:      lastAcceptedByNodeID,
+			getTime:                   getTime,
+		},
+	}
+
+	sd.getSnapshot = sd.getNetworkSnapshot
+	sd.haveWeFailedCatchingUp = sd.failedCatchingUp
+
+	return sd
+}
+
+// CheckIfWeAreStragglingBehind returns for how long our ledger is behind the rest
+// of the nodes in the network. If we are not behind, zero is returned.
+func (sd *stragglerDetector) CheckIfWeAreStragglingBehind() time.Duration {
+	now := sd.getTime()
+	if sd.previousStragglerCheckTime.IsZero() {
+		sd.previousStragglerCheckTime = now
+		return 0
+	}
+
+	// Don't check too often, only once in every minStragglerCheckInterval
+	if sd.previousStragglerCheckTime.Add(sd.minStragglerCheckInterval).After(now) {
+		return 0
+	}
+
+	defer func() {
+		sd.previousStragglerCheckTime = now
+	}()
+
+	if sd.prevSnapshot.isEmpty() {
+		snapshot, ok := sd.getSnapshot()
+		if !ok {
+			sd.log.Trace("No node snapshot obtained")
+			sd.continuousStragglingPeriod = 0
+		}
+		sd.prevSnapshot = snapshot
+	} else {
+		if sd.haveWeFailedCatchingUp(sd.prevSnapshot) {
+			timeSinceLastCheck := now.Sub(sd.previousStragglerCheckTime)
+			sd.continuousStragglingPeriod += timeSinceLastCheck
+		} else {
+			sd.continuousStragglingPeriod = 0
+		}
+		sd.prevSnapshot = snapshot{}
+	}
+
+	return sd.continuousStragglingPeriod
+}
+
+func (sd *stragglerDetector) failedCatchingUp(s snapshot) bool {
+	totalValidatorWeight, nodeWeightsToBlocks := s.totalValidatorWeight, s.nodeWeightsToBlocks
+
+	var processingWeight uint64
+	for nw, lastAccepted := range nodeWeightsToBlocks {
+		if sd.processing(lastAccepted) {
+			newProcessingWeight, err := safemath.Add(processingWeight, nw.Weight)
+			if err != nil {
+				sd.log.Error("Cumulative weight overflow", zap.Uint64("cumulative", processingWeight), zap.Uint64("added", nw.Weight))
+				return false
```

Comment:
should we instead just return err and handle it in either  `CheckIfWeAreStragglingBehind` or again return err from there and handle wherever CheckIfWeAreStragglingBehind is called.

Reply:
This error means the total weight of the subnet has overflowed uint64. I don't think it's worth to change the API to propagate this further up, but if you feel strongly about it, I'll change it accordingly.

Reply:
I'd say we can return err from `failedCatchingUp` and handle in `CheckIfWeAreStragglingBehind`, but really does not super matter if we dont want to return err from `CheckIfWeAreStragglingBehind`

Reply:
the `failedCatchingUp` method (now called `areWeBehindTheRest` ) returns true or false whether it managed to detect the caller node is behind. Handling an error returned by it by the calling method is simply logging it. Therefore I don't see why we need to propagate the error further upwards, if we can handle it at a lower level and simplify the method signature of `failedCatchingUp`.